File: app/src/main/java/br/com/onibusapp/onibusapp/ui/MapsFragment.java
package br.com.onibusapp.onibusapp.ui;

import android.Manifest;
import android.app.Activity;
import android.content.pm.PackageManager;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.ActivityCompat;
import android.support.v4.app.Fragment;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;

import com.google.android.gms.common.ConnectionResult;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.maps.CameraUpdateFactory;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.OnMapReadyCallback;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.MarkerOptions;

import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.TimeUnit;

import br.com.onibusapp.onibusapp.R;
import br.com.onibusapp.onibusapp.ui.mapa.MapsContract;
import br.com.onibusapp.onibusapp.ui.mapa.MapsPresenter;
import br.com.onibusapp.onibusapp.utils.Constantes;

public class MapsFragment extends Fragment implements MapsContract.View, OnMapReadyCallback, GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {

    private GoogleMap mMap;
    private MapsContract.Presenter mMapsPresenter;
    private Timer timer;

   /* @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.fragment_maps);
        // Obtain the SupportMapFragment and get notified when the map is ready to be used.
        SupportMapFragment mapFragment = (SupportMapFragment) getSupportFragmentManager()
                .findFragmentById(R.id.map);
        mapFragment.getMapAsync(this);
        mFusedLocationClient = LocationServices.getFusedLocationProviderClient(this);
    }*/

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_maps, container, false);
        MapView mapView = (MapView) view.findViewById(R.id.mapaFragment);

        mapView.onCreate(savedInstanceState);
        mapView.onResume();
        mapView.getMapAsync(this);
        this.mMapsPresenter = new MapsPresenter(this);
        getParametros();
        return view;
    }

    private void getParametros() {
        Bundle bundle = this.getArguments();
        if (bundle != null) {
            String linha = bundle.getString(Constantes.LINHA);
            Integer sentido = bundle.getInt(Constantes.SENTIDO);
            Integer codigoEmpresa = bundle.getInt(Constantes.CODIGO_EMPRESA);
            Log.d("PARAMENTROS", linha + " " + sentido + " " + codigoEmpresa);
           timer = new Timer();
            timer.scheduleAtFixedRate(new TimerTask() {
                @Override
                public void run() {
                    Log.d("Timer", "EXECUTOU");
                    mMapsPresenter.getLocalizacaoOnibus(linha, sentido, codigoEmpresa);
                }
            },0, TimeUnit.SECONDS.toMillis(30));
        }
    }

    @Override
    public void onMapReady(GoogleMap googleMap) {
        mMap = googleMap;
        mMapsPresenter.mapaPronto();
    }

    @Override
    public void onConnected(@Nullable Bundle bundle) {

    }

    @Override
    public void onConnectionSuspended(int i) {

    }

    @Override
    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {

    }

    @Override
    public void onRequestPermissionsResult(int requestCode, String permissions[], int[] grantResults) {
        switch (requestCode) {
            case Constantes.PERMISSAO_LOCALIZACAO: {
                if (grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    mMapsPresenter.mapaPronto();
                }
                return;
            }
        }
    }

    @Override
    public void mapaPronto() {
        mMap.getUiSettings().setMyLocationButtonEnabled(true);
        mMap.getUiSettings().setMapToolbarEnabled(true);
        if (ActivityCompat.checkSelfPermission(getActivity(), Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                && ActivityCompat.checkSelfPermission(getActivity(), Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(new String[]{ Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION }, Constantes.PERMISSAO_LOCALIZACAO);
            return;
        }
        mMap.setMyLocationEnabled(true);
        this.mMapsPresenter.getLocalizacaoUsuario(true);
    }

    @Override
    public void setLocalizacao(Double lat, Double lnt, Boolean focarLocalizacao) {
        if (mMap != null) {
            LatLng minhaLocalizacao = new LatLng(lat, lnt);
            if (focarLocalizacao) {
                mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(minhaLocalizacao, 14));
                mMap.moveCamera(CameraUpdateFactory.newLatLng(minhaLocalizacao));
            }
        }
    }

    @Override
    public Activity getCurrentActivity() {
        return this.getActivity();
    }

    @Override
    public void limparMapa() {
        mMap.clear();
    }

    @Override
    public void addMarker(LatLng latLng, String linha, String prefixo) {
        mMap.addMarker(new MarkerOptions().position(latLng).title(linha + " - " + prefixo));
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        if (timer != null) {
            timer.cancel();
            timer.purge();
        }
    }

    @Override
    public void onPause() {
        super.onPause();
    }

    @Override
    public void onResume() {
        super.onResume();
    }
}

File: app/src/main/java/br/com/onibusapp/onibusapp/ui/mapa/MapsFragment.kt
package br.com.onibusapp.onibusapp.ui

import android.Manifest
import android.app.Activity
import android.content.pm.PackageManager
import android.os.Bundle
import android.support.design.widget.BottomNavigationView
import android.support.v4.app.ActivityCompat
import android.support.v4.app.Fragment
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import br.com.onibusapp.onibusapp.R
import br.com.onibusapp.onibusapp.ui.mapa.MapsContract
import br.com.onibusapp.onibusapp.ui.mapa.MapsPresenter
import br.com.onibusapp.onibusapp.utils.Constantes
import br.com.onibusapp.onibusapp.utils.ContadorAtualizacao
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.maps.CameraUpdateFactory
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.OnMapReadyCallback
import com.google.android.gms.maps.model.BitmapDescriptorFactory
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.MarkerOptions
import java.util.*
import java.util.concurrent.TimeUnit

class MapsFragment : Fragment(), MapsContract.View, OnMapReadyCallback, GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener {

    private var mMap: GoogleMap? = null
    private var tvLinha: TextView? = null
    private var tvTempoAtualizaca: TextView? = null
    private var progressBar: ProgressBar? = null
    private var mMapsPresenter: MapsContract.Presenter? = null
    private var timer: Timer? = null
    private var contadorAtualizacao: ContadorAtualizacao? = null

    override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        val view = inflater!!.inflate(R.layout.fragment_maps, container, false)
        val mapView = view.findViewById<View>(R.id.mapaFragment) as MapView

        mapView.onCreate(savedInstanceState)
        mapView.onResume()
        mapView.getMapAsync(this)

        val layoutDadosAtualizacao = view.findViewById<View>(R.id.dadosAtualizacao)
        layoutDadosAtualizacao.bringToFront()

        val layoutDadosLinha = view.findViewById<View>(R.id.dadosLinha)
        layoutDadosLinha.bringToFront()

        tvLinha = view.findViewById<View>(R.id.tv_linha) as TextView
        tvTempoAtualizaca = view.findViewById<View>(R.id.tv_tempo_atualizacao) as TextView
        progressBar = view.findViewById<View>(R.id.progress_atualizacao) as ProgressBar

        this.mMapsPresenter = MapsPresenter(this)
        getParametros()
        val navigation = activity.findViewById<View>(R.id.navigation) as BottomNavigationView
        navigation.menu.getItem(0).isChecked = true
        return view
    }

    private fun getParametros() {
        val bundle = this.arguments
        if (bundle != null) {
            val linha = bundle.getString(Constantes.LINHA)
            val sentido = bundle.getInt(Constantes.SENTIDO)
            val url = bundle.getString(Constantes.URL)
            Log.d("PARAMENTROS", "$linha $sentido $url")
            tvLinha!!.setText(linha)
            timer = Timer()
            contadorAtualizacao = ContadorAtualizacao(tvTempoAtualizaca!!, progressBar!!, 20000, 1000)
            timer!!.scheduleAtFixedRate(object : TimerTask() {
                override fun run() {

                    contadorAtualizacao!!.iniciar()

                    Log.d("Timer", "EXECUTOU")
                    mMapsPresenter!!.getLocalizacaoOnibus(linha, sentido, url)
                    //contadorAtualizacao!!.onFinish()
                    /*contadorAtualizacao = ContadorAtualizacao(tvTempoAtualizaca!!, 20000, 1000, true)
                    contadorAtualizacao!!.create()*/

                    //tempoAtualizacao();
                }
            }, 0, TimeUnit.SECONDS.toMillis(20))
        }
    }

    override fun onMapReady(googleMap: GoogleMap) {
        mMap = googleMap
        mMapsPresenter!!.mapaPronto()
    }

    override fun onConnected(bundle: Bundle?) {

    }

    override fun onConnectionSuspended(i: Int) {

    }

    override fun onConnectionFailed(connectionResult: ConnectionResult) {

    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<String>, grantResults: IntArray) {
        when (requestCode) {
            Constantes.PERMISSAO_LOCALIZACAO -> {
                if (grantResults.size > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
                    mMapsPresenter!!.mapaPronto()
                }
                return
            }
        }
    }

    override fun mapaPronto() {
        mMap!!.uiSettings.isMyLocationButtonEnabled = true
        mMap!!.uiSettings.isMapToolbarEnabled = true
        if (ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED && ActivityCompat.checkSelfPermission(activity, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            requestPermissions(arrayOf(Manifest.permission.ACCESS_FINE_LOCATION, Manifest.permission.ACCESS_COARSE_LOCATION), Constantes.PERMISSAO_LOCALIZACAO)
            return
        }
        mMap!!.isMyLocationEnabled = true
        this.mMapsPresenter!!.getLocalizacaoUsuario(true)
    }

    override fun setLocalizacao(lat: Double?, lnt: Double?, focarLocalizacao: Boolean?) {
        if (mMap != null) {
            val minhaLocalizacao = LatLng(lat!!, lnt!!)
            if (focarLocalizacao!!) {
                mMap!!.moveCamera(CameraUpdateFactory.newLatLngZoom(minhaLocalizacao, 14f))
                mMap!!.moveCamera(CameraUpdateFactory.newLatLng(minhaLocalizacao))
            }
        }
    }

    override fun getCurrentActivity(): Activity {
        return this.activity
    }



    override fun limparMapa() {
        mMap!!.clear()
    }

    override fun addMarker(latLng: LatLng, linha: String, prefixo: String) {
        mMap!!.addMarker(MarkerOptions().position(latLng).title("$linha - $prefixo")).setIcon(BitmapDescriptorFactory.fromResource(R.drawable.round_directions_bus_black_18dp))
    }

    override fun onDestroy() {
        super.onDestroy()
        this.limparMapa()

        if (timer != null) {
            timer!!.cancel()
            timer!!.purge()
        }
        mMapsPresenter = null

        if (contadorAtualizacao != null) {
            this.contadorAtualizacao!!.onFinish()
        }

    }

}
